pkg/controllers/multicluster/cloudmap_controller.go

```diff
@@ -131,41 +131,62 @@ func (r *CloudMapReconciler) reconcileService(ctx context.Context, svc *model.Se
 
 	importedSvcPorts := ExtractServicePorts(svc.Endpoints)
 
+	clusterIdToEndpointsMap := make(map[string][]*model.Endpoint)
+	for _, ep := range svc.Endpoints {
+		clusterIdToEndpointsMap[ep.ClusterId] = append(clusterIdToEndpointsMap[ep.ClusterId], ep)
+	}
+
+	clusterIds := make([]string, 0, len(clusterIdToEndpointsMap))
+	for clusterId := range clusterIdToEndpointsMap {
+		clusterIds = append(clusterIds, clusterId)
+	}
+
 	svcImport, err := r.getServiceImport(ctx, svc.Namespace, svc.Name)
 	if err != nil {
 		if !errors.IsNotFound(err) {
 			return err
 		}
 
 		// create ServiceImport if it doesn't exist
-		if svcImport, err = r.createAndGetServiceImport(ctx, svc.Namespace, svc.Name, importedSvcPorts); err != nil {
+		if svcImport, err = r.createAndGetServiceImport(ctx, svc.Namespace, svc.Name, importedSvcPorts, clusterIds); err != nil {
 			return err
 		}
 	}
 
-	derivedService, err := r.getDerivedService(ctx, svc.Namespace, svcImport.Annotations[DerivedServiceAnnotation])
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			return err
+	// get or create derived Service for each cluster the service is a member of
+	derivedServices := make([]*v1.Service, 0, len(clusterIdToEndpointsMap))
+	for _, clusterId := range clusterIds {
+		endpoints := clusterIdToEndpointsMap[clusterId]
+		clusterImportedSvcPorts := ExtractServicePorts(endpoints)
+
+		derivedService, err := r.getDerivedService(ctx, svc.Namespace, svc.Name, clusterId)
```

Comment:
We are not covering for the scenario when the service is removed from one of the cluster. In that case we should remove the derived service of that cluster. Best way to do this would be to fetch list of derived services, and delete those without any endpoints in cloudmap.